pkg/kubelet/cm/deviceplugin/pod_devices.go

```diff
@@ -17,241 +17,507 @@ limitations under the License.
-	if !exists {
-		return
+// recordReservations records pre-allocation results into registry.
+func (a *devAllocationImpl) recordReservations(req *allocRequest, selected map[string]sets.String) {
+	// copy device list
+	devices := make(map[string][]string)
+	for r, s := range selected {
+		devices[r] = s.List()
 	}
-	resources, exists := containers[contName]
-	if !exists {
-		return
+
+	// devices are reusable across init containers
+	for container, rmap := range req.inits {
+		for resource, count := range rmap {
+			if count > 0 {
+				devList := devices[resource][:count]
+				assigned := sets.NewString(devList...)
+				a.insert(req.pod, container, resource, assigned, nil)
+			}
+		}
 	}
-	for resource, devices := range resources {
-		allocatedResources[resource] = allocatedResources[resource].Union(devices.deviceIds)
+
+	for container, rmap := range req.containers {
+		for resource, count := range rmap {
+			if count > 0 {
+				devList := devices[resource][:count]
+				assigned := sets.NewString(devList...)
+				a.insert(req.pod, container, resource, assigned, nil)
+				devices[resource] = devices[resource][count:]
+			}
+		}
 	}
 }
 
-// Removes the device resources allocated to the specified <podUID, contName> from allocatedResources.
-func (pdev podDevices) removeContainerAllocatedResources(podUID, contName string, allocatedResources map[string]sets.String) {
-	containers, exists := pdev[podUID]
-	if !exists {
-		return
-	}
-	resources, exists := containers[contName]
-	if !exists {
-		return
-	}
-	for resource, devices := range resources {
-		allocatedResources[resource] = allocatedResources[resource].Difference(devices.deviceIds)
+// confirm changes all device reservations into allocated status
+func (a *devAllocationImpl) confirm(req *allocRequest, responses map[string]*pluginapi.AllocateResponse) error {
+	a.Lock()
+	defer a.Unlock()
+
+	// devices are reusable across init containers
+	for container, rmap := range req.inits {
+		for resource, count := range rmap {
+			if count > 0 {
+				old := a.registry[req.pod][container][resource]
+				a.registry[req.pod][container][resource] = deviceInfo{
+					status:    devAllocated,
+					ids:       old.ids,
+					allocResp: responses[resource],
+				}
+			}
+		}
 	}
-}
 
-// Returns all of devices allocated to the pods being tracked, keyed by resourceName.
-func (pdev podDevices) devices() map[string]sets.String {
-	ret := make(map[string]sets.String)
-	for _, containerDevices := range pdev {
-		for _, resources := range containerDevices {
-			for resource, devices := range resources {
-				if _, exists := ret[resource]; !exists {
-					ret[resource] = sets.NewString()
+	for container, rmap := range req.containers {
+		for resource, count := range rmap {
+			if count > 0 {
+				old := a.registry[req.pod][container][resource]
+				a.registry[req.pod][container][resource] = deviceInfo{
+					status:    devAllocated,
+					ids:       old.ids,
+					allocResp: responses[resource],
```

Comment:
As device is not considered to get reused between regular containers, I think we could not share a same response here.

Reply:
As commented above, devices are not reused across regular containers.

Reply:
But what I could see from the change is that we do allocation for a pod only once(i.e allocation per pod), then in `collectRunOptions`, all of its containers fetch runtime info from a same response.

Let's skip the issue of reuse across regular containers temporarily, say we have two regular containers, each of which requests 1 GPU, as a result, we allocate GPU A and B for them, but when collecting run options, both of them would  get runtime info of the two GPUs, thus, we could not ensure they each use a different GPU, and may lead to a conflict as a result

Reply:
Take the GPU as an example, a single resource type, we issue a single call to the endpoint (according to current gRPC design), see the following line of code:

```
func (m *ManagerImpl) Allocate() {
   // ...
  for resource, devs := range selected {
    resp, err := ep.allocate(devs.List())
    responses[resource] = resp
  }
```

This invocation means the device allocation response can be reused across all the allocated devices.

Reply:
Anyway, if we believe that the allocate call should be performed on a per-container basis, this code can be easily changed.

Reply:
But I think a device could not be used by multi containers at the same time

Reply:
No. Devices are NOT reused across regular containers.

Reply:
Yep, we did not do reused across regular containers during allocation, what I'm concerned about is that when collecting run options, we does not generate seperate runtime info for the containers, they both could see the full information of all devices.

Reply:
@lichuqiang i am not sure that i understood your concern. While collecting runtime info, registry will be looked into and key to find resource info is pod+container. Therefore, i am unable to understand that how a regular container can access resource info of the resource allocated to another regular container. Can you please try to reword your concern.

Reply:
>While collecting runtime info, registry will be looked into and key to find resource info is pod+container

The problem is that the response of all the containers in a pod are the same, i.e. the response we got during allocation

Reply:
Which is coherent with the current behavior right ? And the model is extensible so we can support per container allocation for 1.10.

Given the following spec:
- InitContainer (foo) requests 1 GPU
    - Is assigned GPU A
- Container (bar) requests 2 GPUs
    - Is assigned GPU A and B

Previous behavior was to issue at the "same time" (pod creation):
    - 1 call to Allocate(A)
    - 1 call to Allocate(A, B)

Current behavior issues:
    - 1 call to Allocate(A, B)

The 2nd behavior seems more consistent with pod level device isolation.
The first behavior is wrong and has one foot in both per container isolation and per pod isolation 's doors. Additionally no known device plugin currently does anything special with that behavior.

Reply:
Yep, there is a problem as @RenaudWasTaken rightly explained. This also suggests that we need to add some more unit tests to catch this type of scenario.
@tengqm looks like allocate call per-container is required.

Reply:
>Which is coherent with the current expected behavior right ? 

Nope, it's okay for a init container and a regular, but not for two regular containers.
Since a device could not be used by multi containers at the same time, for regular containers, we need to provide runtime information telling the container which device to use, a shared runtime information could lead to conflict.

Reply:
I don't think issuing a single call to allocate(A, B) is the correct behavior. It could generate wrong runtime configuration for the container that just requesting GPU A.